### DISCUSSION TITLE
Having a lot of memory issues on my server

### DISCUSSION THREAD

**Discussion by SteveSharpe** [mbround18/valheim-docker#328] [Help!] [answered]
I'm running an Ubuntu server in AWS Lightsail using mbround18's docker image for a dedicated server. The instance I'm using has 4 GB of RAM, which was plenty for the first month or so of use. Lately the server has been locking up a lot. I can only keep it online maybe 4-5 hours before it will lock up again.

I started keeping an eye on what is causing it and I've basically narrowed it down to RAM getting maxed out on the server. When valheim_server first starts, it consumes about 2 GB of RAM. Sometimes a little higher than that. I don't see any noticeable increase in RAM consumption even when 5 or 6 people are on the server. However, after some amount of time goes by, the RAM will just slowly creep up to 2.5, 3, 3.5... once it gets to about 3.5 the server is fully maxed and either the valheim_server process will get killed or the server just completely hangs and I have to reboot.

Today I did an experiment where I rebooted the server and started the valheim_server process and kept an eye on memory utilization while there were zero clients connected. No one joined the server at all while I was watching, but the memory still kept going up until eventually the server crashed.

I feel like this has only started happening since the last Valheim patch.  I doubt it is due to the container, but I can't be sure.  I'm just curious if anyone else is having this kind of memory utilization issue?

I'd rather not just bump the RAM on the instance and pay more money to solve this issue.  But if I have to and I'm not the only one with this issue, mbround18 would probably need to update the getting started guide to recommend something more than 2 GB RAM for this server.  It seems to be pretty RAM hungry these days.

**Comment by mbround18** [comment, marked as answer]
Hey @SteveSharpe, are you running any mods by chance? I am running a vanilla server and a ValheimPlus server which virtually have no limits on ram. Both servers are hovering around 2.9-3.3 gigs of ram idle and with 4-5 online, it hovers around 4 gigs. 

![image](https://user-images.githubusercontent.com/12646562/114485230-6aa29c80-9bc0-11eb-8b54-f98c13525cae.png)

I just gave it a reboot (thats the dip toward current), once its back online ill monitor it for a bit and see if I run into a similar issue.

**Reply by SteveSharpe** [reply]
Thanks @mbround18.  I'm not running any mods, not even ValheimPlus.  Everything is vanilla.  The only thing I've even changed with your container is I added a "restart: unless-stopped" to the docker-compose so that when I reboot the server (which I now have to do a lot), the container will come back up automatically.

One thing that's interesting with what I'm seeing so far is that the memory usage leak actually seems worse when no players are connected.  I am able to play for hours with sometimes 5 or 6 people on the server and the memory utilization will hover around 3 GB out of 4 GB available.  But if everyone disconnects, the RAM utilization will slowly creep up until the server crashes or oom-killer destroys the valheim_server process.  It takes about 2 hours now for the server to crash, with no players connected at all.

**Reply by mbround18** [reply]
huh, ill give this a whirl on a test server :P

**Reply by SteveSharpe** [reply]
Did you learn anything?  I think it's probably safe to say that in the current state, 8 GB is a safer RAM amount than 4 GB or lower.

**Reply by mbround18** [reply]
After some research, I have encountered the memory increasing only once. This was after the peeps on my server did some massive terraforming. I imagine its the server indexing all the new updates to the landscape and saving them to the world data base.

**Reply by SteveSharpe** [reply]
Whatever is causing the memory leak, it only seems to do it on my server when no one is connected.  Yesterday there were several of us connected all day long, coming and going, and the memory stayed at a pretty steady 3 GB utilization.  Definitely higher than when the server first started and before we did a lot of terraforming, but while at least one of us was connected, the server stayed steadily at that 3 GB.

What happens now is that about 2 hours (almost exactly 2 hours) of running with no players connected, the server will max memory and crash.  If I reboot the server, start Valheim, and let it sit with no players--it will crash in 2 hours.  If I start the server and play on it for an hour and then disconnect--it will crash 2 hours after I log out.  Yesterday we had the server running with us playing for about 6 hours.  From 6pm to midnight.  When I looked into the logs this morning, it crashed at 2am after we had logged out.

Nothing you or I can do about it, but it's interesting nonetheless.

**Reply by robcz** [reply]
I'm facing a similar issue, which appears to have at least coincidentally cropped up after the 0.150.3 update.  If players are on the server, everything operates as normal.  Once players are off of the server, after about 2 hours it becomes non-responsive to the point that I can't even get an ssh session into the server.  I've also noticed that it is now during that idle state time burning heavily through burst cpu.  No-mods on my server.

**Reply by robcz** [reply]
> I'm facing a similar issue, which appears to have at least coincidentally cropped up after the 0.150.3 update. If players are on the server, everything operates as normal. Once players are off of the server, after about 2 hours it becomes non-responsive to the point that I can't even get an ssh session into the server. I've also noticed that it is now during that idle state time burning heavily through burst cpu. No-mods on my server.

I bumped up to the 4 vCPUs, 16 GB memory instance and the crashes stopped.  I'll ratchet it down to a smaller instance and see if it starts happening again.

**Reply by mikegreen** [reply]
Just checked mine - when this issue was logged, I was on an older version and even with 4 players it'd cap out at 2.2-2.5gb of ram used by the container.  Upgraded about a week ago? Now, seeing it creep up to 3.1gb after 8hrs of idle.

**Reply by SteveSharpe** [reply]
For me I believe this started happening the version prior to 0.150.3.  It is definitely still happening since the latest patch. 

I haven’t upgraded the RAM on my server, but others have said that with 8 GB it doesn’t seem to run out. 

I noticed something when I had htop running in a window to the side while me and some friends were playing the other day. While we are connected, the server will bounce between 2.6 and 3.1 GB RAM utilization. It actually goes up and sometimes down. As soon as everyone is disconnected, it will go up to the 3.1, then 3.5, and eventually it steps up too high and the server runs out. 

It’s almost like there’s some cleanup that is running properly when players are connected, but runs out of control when everyone is logged off. Perhaps more RAM isn’t the fix, but basically just gives more time for the server to slowly consume more RAM until the next player logs in. I might set up a local server to test it out. For now I don’t want to double my AWS cost, which is what it’ll take to upgrade from 4 to 8 GB.

**Reply by guidodavide** [reply]
My 2 cents.

I can't speak of RAM usage, since I've never faced this issue, but you are saying that your server runs without players being connected for many hours.

Why don't you shut it down and start only when needed? That will also help you saving some money on AWS.

I've my server running on Google Cloud and I set up the start/stop instance permission for all my friends. They are simply starting the instance/stopping it when needed and check whether someone is connected with the HTTP status server.
(Discord webhooks help too to monitor the state, which I customized a bit).

It's a workaround, but it will help you saving money and avoiding the out of memory issue.
(I have a 2VCPU(shared cores) 4GB instance)

**Reply by SteveSharpe** [reply]
I’ve built my VM in Lightsail, which has a fixed cost per month. Doesn’t matter if it’s up or down. At this point we’re just going to finish the last boss and then I’ll delete the server. Hopefully server performance issues get better as they develop the game.
